teamcity.build

```diff
@@ -245,231 +246,278 @@
-										missingTests.Add(result);
-
-								report.AppendLine();
-								report.AppendLine("*** Tests missing since last recorded results ***");
-								if (missingTests.Count > 0)
-								{
-									foreach(Result result in missingTests)
-										report.AppendLine(result.ToString());
-								}
-								else
-									report.AppendLine("None");
-									
-								IList<Result> ignoredTests = new List<Result>();
-								foreach(Result afterResult in afterExistingTests)
-									foreach(Result beforeResult in before)
-										if (beforeResult.Name == afterResult.Name)
-											if (beforeResult.Executed && !afterResult.Executed)
-												ignoredTests.Add(afterResult);
-
-								report.AppendLine();
-								report.AppendLine("*** Tests ignored since last recorded results ***");
-								if (ignoredTests.Count > 0)
-								{
-									foreach(Result result in ignoredTests)
-										report.AppendLine(result.ToString());
-								}
-								else
-									report.AppendLine("None");
-
-								IList<Result> brokenTests = new List<Result>();
-								foreach(Result afterResult in afterExistingTests)
-									foreach(Result beforeResult in before)
-										if (beforeResult.Name == afterResult.Name)
-											if (beforeResult.Success && afterResult.Executed && !afterResult.Success)
-												brokenTests.Add(afterResult);
-
-								report.AppendLine();
-								report.AppendLine("*** Tests broken since last recorded results ***");
-								if (brokenTests.Count > 0)
-								{
-									foreach(Result result in brokenTests)
-										report.AppendLine(result.ToString());
-								}
-								else
-									report.AppendLine("None");
-
-								string output = report.ToString();
-								File.WriteAllText(outputFile, output);
-							
-								if (brokenTests.Count > 0)
-									throw new Exception("Previously passing tests have been broken\n\n" + output);
-							
-								if (newFailingTests.Count > 0)
-									throw new Exception("New tests that fail have been added\n\n" + output);
-							
-								return output;
-							}
-							catch(Exception e)
-							{
-								Console.WriteLine(e);
-								throw;
-							}
-						}
-
-            ]]>
+						else
+							newNotFailingTests.Add(afterResult);
+					}
+				}
+
+				var afterTestNames = new HashSet<string>();
+				foreach (var result in after) afterTestNames.Add(result.Name);
+				var missingTests = new List<Result>();
+				foreach (var result in before)
+					if (!afterTestNames.Contains(result.Name))
+						missingTests.Add(result);
+
+				report.AppendLine();
+				report.AppendLine("*** Tests broken since last recorded results ***");
+				if (brokenTests.Count > 0)
+				{
+					foreach (var result in brokenTests)
+						report.AppendLine(result.ToString());
+				}
+				else
+					report.AppendLine("None");
+
+				report.AppendLine();
+				report.AppendLine("*** Tests new (failed) since last recorded results ***");
```

Comment:
I have moved the failed tests at start of file. Seeking them in the middle of other was not very practical.